### ISSUE TITLE
Handle `Incoming::PubAck` in the `mqqt` source

### ISSUE THREAD

**Issue by pront** [vectordotdev/vector#21967]
### A note for the community

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

Ref: https://github.com/vectordotdev/vector/issues/584

**Comment by antaresindustries** [comment]
We'd like to pick this up and propose a concrete scope, since the fields above are
still stubbed. We have a working implementation with tests and intend to open a PR.

**Scope clarification.** The underlying goal here is **end-to-end acknowledgement
support for the `mqtt` source** (so it can provide at-least-once delivery). For a
source — which only *subscribes* — the mechanism is to defer the source's
**outbound** QoS-1 `PUBACK` until the event has been delivered downstream, using
rumqttc's manual-ack mode. (`Incoming::PubAck` itself only arrives if the client
*publishes* at QoS 1, which the source never does, so handling that packet isn't
the lever for a subscribe-only source.)

### Use Cases

Ingesting telemetry published to an MQTT broker at QoS 1 with a persistent session
(`clean_session = false`), where the goal is **no data loss if Vector restarts,
crashes, or a sink is briefly unavailable**. The broker already provides the
durability primitive — it redelivers any message the client hasn't acknowledged —
but only if the consuming client defers its `PUBACK` until the message is durably
handled.

Today the `mqtt` source reports `can_acknowledge() = false` and auto-acks every
incoming publish on receipt (before the event reaches, let alone is accepted by, a
sink). Any event in flight when Vector stops or a sink write fails is silently
lost, even though the broker had a durable copy it would have redelivered. This
blocks replacing a bespoke ingester (QoS 1 + manual acks) with Vector.

### Attempted Solutions

Enabling end-to-end acknowledgements at the sink, with the source on a persistent
session, has no effect — because `can_acknowledge()` is `false`, the topology never
wires the source into the ack chain, and the source has already auto-acked the
broker:

```yaml
sources:
  mqtt_in:
    type: mqtt
    host: broker
    topic: "telemetry/#"
    decoding:
      codec: bytes
sinks:
  out:
    type: influxdb_logs
    inputs: [mqtt_in]
    acknowledgements:
      enabled: true
```

Killing Vector mid-write (sink destination unavailable) loses the in-flight
message; the broker considers it delivered and does not redeliver. Disk buffers
don't help — the loss is at the source edge, before any Vector buffer.

### Proposal

Make the `mqtt` source a participating source, following the `kafka`/`aws_sqs`
pattern (ack only after `BatchStatus::Delivered`):

- Add an `acknowledgements` (`SourceAcknowledgementsConfig`) option; `can_acknowledge()` returns `true`.
- When enabled, set `MqttOptions::set_manual_acks(true)` (supported by the pinned `rumqttc`).
- Attach a `BatchNotifier` to decoded events; register each publish with an `OrderedFinalizer`; send `client.ack(&publish)` only on `BatchStatus::Delivered`. On `Errored`/`Rejected`, don't ack — with the existing `clean_session = false` + QoS `AtLeastOnce`, the broker redelivers after reconnect.

Default behavior is unchanged (opt-in). Usage is then `acknowledgements.enabled: true` on the sink (or globally).

We have this implemented with unit + integration tests (an integration test asserts
the broker redelivers an unacknowledged message) and have validated the full path
end-to-end against a real broker + sink. Happy to open the PR against this issue.

### Related — the `nats` source

The `nats` source reports `can_acknowledge() = true` but, for JetStream, calls
`msg.ack()` immediately after decode succeeds (`run_nats_jetstream`) — i.e. before
the sinks confirm delivery. That's the same class of problem (and arguably worse,
since the capability is advertised but not actually end-to-end). It likely warrants
a sibling issue with the same fix shape (`BatchNotifier` + finalizer; ack on
`Delivered`). Happy to file that separately if useful.

**Comment by pront** [comment]
Hi @antaresindustries, thank you for your interest in Vector. The suggested implementation plan makes sense.

> It likely warrants a sibling issue with the same fix shape (BatchNotifier + finalizer; ack on
Delivered). Happy to file that separately if useful.

Yes please create a separate issue.
